cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaMetadataStateManager.java

```diff
@@ -0,0 +1,397 @@
+     */
+    public static final String DISABLED_VALUE_STRIMZI_IO_KRAFT = "disabled";
+
+    private final Reconciliation reconciliation;
+
+    private KafkaMetadataState metadataState;
+
+    private String kraftAnno;
+
+    private KafkaAgentClient kafkaAgentClient;
+    private boolean isMigrationDone = false;
+
+    /**
+     * Constructor
+     *
+     * @param reconciliation Reconciliation information
+     * @param kafkaCr instance of the Kafka CR
+     */
+    public KafkaMetadataStateManager(
+            Reconciliation reconciliation,
+            Kafka kafkaCr) {
+        this.reconciliation = reconciliation;
+        this.kraftAnno = kraftAnnotation(kafkaCr);
+        String metadataStateFromCr = kafkaCr.getStatus() != null ? kafkaCr.getStatus().getKafkaMetadataState() : null;
+        // missing means reconciling an already existing Kafka resource with newer operator supporting metadata state or first reconcile
+        if (metadataStateFromCr == null) {
+            this.metadataState = isKRaftEnabled(this.kraftAnno) ? KafkaMetadataState.KRaft : KafkaMetadataState.ZooKeeper;
+        } else {
+            this.metadataState = KafkaMetadataState.valueOf(metadataStateFromCr);
+        }
+        LOGGER.infoCr(reconciliation, "Loaded metadata state from the Kafka CR [{}] and strimzi.io/kraft [{}]", this.metadataState, this.kraftAnno);
+    }
+
+    /**
+     * Computes the next state in the metadata migration Finite State Machine (FSM)
+     * based on the current state and the strimzi.io/kraft annotation value at the
+     * beginning of the reconciliation when this instance is created
+     *
+     * @return the next FSM state
+     */
+    public KafkaMetadataState computeNextMetadataState() {
+        KafkaMetadataState currentState = this.metadataState;
+        String currentKraftAnno = this.kraftAnno;
+        this.metadataState = switch (currentState) {
+            case KRaft -> onKRaft(reconciliation, currentKraftAnno);
+            case ZooKeeper -> onZooKeeper(reconciliation, currentKraftAnno);
+            case KRaftMigration -> onKRaftMigration(reconciliation, currentKraftAnno);
+            case KRaftDualWriting -> onKRaftDualWriting(reconciliation, currentKraftAnno);
+            case KRaftPostMigration -> onKRaftPostMigration(reconciliation, currentKraftAnno);
+        };
+        LOGGER.infoCr(reconciliation, "from [{}] to [{}] with strimzi.io/kraft: [{}]", currentState, this.metadataState, currentKraftAnno);
+        return this.metadataState;
+    }
+
+    /**
+     * Compute the next desired nodes configuration related state
+     * Starting from the current internal FSM state and taking into account the strimzi.io/kraft annotation value,
+     * it computes a state representing the desired configuration for nodes (brokers and/or controllers)
+     *
+     * @return the next desired nodes configuration related state
+     */
+    public KafkaMetadataConfigurationState computeMetadataConfigurationState() {
```

Comment:
Actually it has a different meaning ...

The `computeNextMetadataState` is really about computing the next state of the internal FSM from the current one.
The `computeMetadataConfigurationState` is about getting the metadata configuration state from the current internal FSM state, so it's not moving the FSM through states. I was wondering if it should be more a `get` instead of a `compute`.

Reply:
Ok, I think "get" is better.